### PR TITLE
Feature/datetime functions

### PR DESCRIPTION
**Pull request by Yuhuishishishi**
Noob here.

I am trying to help with #11.
I have done a few remaining datetime related functions:
* TIMESTAMPADD
* LASTDAY
* CEIL (time unit=YEAR, WEEK, MONTH, QUARTER currently do not work)
* FLOOR (time unit=YEAR, WEEK, MONTH, QUARTER currently do not work)

I am still working on:
* TIMESTAMPDIFF

Would mind helping me review the diff and provide a few directions? Thanks

### CODE REVIEW

**Comment by nils-braun** [comment]
Thank you so much @Yuhuishishishi and a very warm welcome to the project! Happy to have you with me.

That is definitely not a "noob" PR!.  Just some general thoughts (which you have probably already thought of), before I add some detailed comments: 
* could you fix the style errors so that we have a consistent style? Running "black" on the files should probably do the trick.
* do you plan to add some tests or should I help you with this?

I guess you are planning to do these things anyway, just to have them written somewhere.

File: dask_sql/mappings.py [hunk 1]
```diff
@@ -106,10 +106,26 @@ def sql_to_python_value(sql_type: str, literal_value: Any) -> Any:
...
+                return delta
+        except IndexError:
+            # no finer granular interval type specified
+            pass
+        except TypeError:
+            # interval type is not recognized, fall back to default case
+            pass
+
         # Calcite will always convert to milliseconds
         # no matter what the actual interval is
```
Comment (nils-braun) [inline dask_sql/mappings.py:125]: That seems to not be correct, or? Maybe we should change the comment then.
Thanks for finding out!

File: dask_sql/mappings.py [hunk 2]
```diff
@@ -106,10 +106,26 @@ def sql_to_python_value(sql_type: str, literal_value: Any) -> Any:
...
         # I am not sure if this breaks somewhere,
         # but so far it works
+        # Issue: if sql_type is INTERVAL MICROSECOND, and value <= 1000, literal_value will be rounded to 0
         return timedelta(milliseconds=float(str(literal_value)))
```
Comment (nils-braun) [inline dask_sql/mappings.py:129]: What is the benefit of your class (`pd.tseries.offsets.DateOffset`) to the `datetime` timedelta? Just for me to understand - I am very fine with changing it

File: dask_sql/physical/rex/core/call.py [hunk 3]
```diff
@@ -416,6 +422,48 @@ def extract(self, what, df: SeriesOrScalar):
...
+    Apply ceil/floor operations on a series depending on its dtype (datetime like vs normal)
+    """
+
+    def __init__(self, round_method: str):
+        assert round_method in {
+            "ceil", "floor"
+        }, f"Round method can only be either ceil or floor"
+
+        super().__init__(
+            lambda x: pd.api.types.is_datetime64_any_dtype(x) or isinstance(x, datetime), # if the series is dt type
```
Comment (nils-braun) [inline dask_sql/physical/rex/core/call.py:436]: Maybe create a temporary variable for the lambda function

File: dask_sql/physical/rex/core/call.py [hunk 4]
```diff
@@ -416,6 +422,48 @@ def extract(self, what, df: SeriesOrScalar):
...
+            self._round_datetime,
+            getattr(da, round_method)
+        )
+
+        self.round_method = round_method
+
+    def _round_datetime(self, *operands):
+        df, unit = operands
+
+        if is_frame(df):
```
Comment (nils-braun) [inline dask_sql/physical/rex/core/call.py:446]: As we need this part multiple times (here and e.g. also in the "last_day" function): should we create a utility function?

File: dask_sql/physical/rex/core/call.py [hunk 5]
```diff
@@ -37,26 +37,32 @@ def of(self, op: "Operation") -> "Operation":
         return Operation(lambda x: self(op(x)))
 
 
-class TensorScalarOperation(Operation):
+class PredicteBasedOperation(Operation):
```
Comment (nils-braun) [inline dask_sql/physical/rex/core/call.py:40]: Nice one!

**Comment by Yuhuishishishi** [comment]
> * could you fix the style errors so that we have a consistent style? Running "black" on the files should probably do the trick.
The `black` command triggers reformatting some of the lines that are not originally included in this PR, thus creating some additional fmt only line diffs. Is it fine to do so?

> do you plan to add some tests for the remaining uncovered lines or should I help you with this?
I am pretty new Codecov, but I think I am going to figure out the uncovered lines and will add tests to cover them (mostly exception handling logics)

File: dask_sql/mappings.py (hunk 2 above)
Comment (Yuhuishishishi) [reply dask_sql/mappings.py:129]: With `pd.tseries.offsets`, my understanding is that it's more convenient to specify timedelta with data-dependent length. E.g., when incrementing the current timestamp by "1 month", with standard timedelta, we need to compute if to add 30/31/28/29 days depending the month/year of the current timestamp. 
With `pd.tseries.offsets`, we can use `pd.tseries.offsets.DateOffset(months=1), and pandas shall take the day conversion under the hood.

Similar thing happens to "year" increment as well due to leap years.

For `last_day` implementation, it is convenient to use  `pd.tseries.offsets.MonthEnd`, which shifts the time to the last day of the month.

For more generic cases (such as + hours/minutes/seconds), I think the standard timedelta and `pd.tseries.offsets` will all do the tricks.

**Comment by nils-braun** [comment]
> The `black` command triggers reformatting some of the lines that are not originally included in this PR, thus creating some additional fmt only line diffs. Is it fine to do so?

Hm, that is strange (because lines with wrong formatting should not be committable). Which version of black are you using? I am using `black, version 19.10b0`.

> I am pretty new Codecov, but I think I am going to figure out the uncovered lines and will add tests to cover them (mostly exception handling logics). Adding a "WIP" mark to the PR before they are completed.

Probably you already know, but you can also run the tests locally with coverage display:
```
pytest tests --cov-report term-missing
```

But if you need any help, I am happy to help you find out!

**Comment by Yuhuishishishi** [comment]
@nils-braun I have updated the PR to address some of your comments. Thanks for all the nice suggestions.

**Comment by nils-braun** [comment]
Very nice @Yuhuishishishi 
That one is ready to merge. If you have any more additions to it, feel free to open a new PR.
Very happy to have you on board in this project!

**Comment by Yuhuishishishi** [comment]
@nils-braun Appreciate all the comments and help. Glad to help with the project!

**Pull request merged**